### DISCUSSION TITLE
I don't understand  proxy-rewrite - regex_uri

### DISCUSSION THREAD

**Discussion by woyouyitouchenfeijv** [apache/apisix#6490] [Q&A]
我有两个Route

第一个直接请求code，然后不动转发到后端 1.1.1.1:8080/code。用的域名是aaa.com.cn
第二个是请求code2，然后会正则变成code，请求后端的2.2.2.2:8080/code，域名也是aaa.com.cn

这个时候，请求code2的时候，apisix会把请求变成code后，请求第一个route，变成请求1.1.1.1:8080；而不是请求route2的上游。如果没有第一个请求，就完全没问题，请求2.2.2.2:8080。

按道理apisix如果认为正则转发后的是一个route时，没有第一个route不应该报错404吗？

En:
I have two routes
The first one directly requests code, and then forwards it to the backend 1.1.1.1:8080 / code. The domain name used is AAA com. cn
The second is to request code2, and then it will be regularized into code. The back-end of the request is 2.2.2.2:8080 / code, and the domain name is AAA com. cn
At this time, when requesting code2, apisik will change the request into code and request the first route into request 1.1.1.1:8080; Instead of requesting the upstream of route2. If there is no first request, there is no problem at all. Request 2.2.2.2:8080.
According to the truth, if apisik thinks that the route after regular forwarding is a route, shouldn't it report Error 404 without the first route?

Route 1

{
  "uri": "/code*",
  "name": "code",
  "host": "aaa.com.cn",
  "upstream": {
    "nodes": [
      {
        "host": "1.1.1.1",
        "port": 8080,
        "weight": 1
      }
    ],
    "type": "roundrobin",
    "scheme": "http",
    "pass_host": "pass",
    "keepalive_pool": {
      "idle_timeout": 60,
      "requests": 1000,
      "size": 320
    }
  },
  "status": 1
}

------------------------------------------------------

Route 2
{
  "uri": "/code2*",
  "name": "code2",
  "host": "aaa.com.cn",
  "plugins": {
    "proxy-rewrite": {
      "regex_uri": [
        "^/code2(.*)",
        "/code$1"
      ]
    }
  },
  "upstream": {
    "nodes": [
      {
        "host": "2.2.2.2",
        "port": 8080,
        "weight": 1
      }
    ],
    "timeout": {
      "connect": 6,
      "send": 6,
      "read": 6
    },
    "type": "roundrobin",
    "scheme": "http",
    "pass_host": "pass",
    "keepalive_pool": {
      "idle_timeout": 60,
      "requests": 1000,
      "size": 320
    }
  },
  "status": 1
}

```

**Comment by tzssangglass** [comment]
> when requesting code2, apisik will change the request into code and request the first route into request 1.1.1.1:8080;

How did you verify that the request was forwarded to 2.2.2.2? Basically this is not possible.

**Reply by tzssangglass** [reply]
APISIX will not reroute the request.

**Reply by woyouyitouchenfeijv** [reply]
no I mean, when there is only the second, the request will arrive at 2.2.2.2.

However, when both the first and second requests exist, why is the second request re requested as a route after being regularly forwarded.

If it is treated as a new route after each regularization, why not report Error 404 when there is only the second route

**Reply by tzssangglass** [reply]
> when both the first and second requests exist, why is the second request re requested as a route after being regularly forwarded.

I don't know what you mean "second request re requested as a route after being regularly forwarded", as I said, when a request enters APISIX, it will only be matched route once.

 Can you provide a complete reproduction of the process? Although you provided the route configuration above, you are still missing the

1. how did you validate it, so far you are only talking about the phenomenon, but not how you validated it?
2. how does your upstream handle the request? For example, how do you handle `/code` requests?

**Reply by tzssangglass** [reply]
I need a clear process like https://github.com/apache/apisix/issues/6095 and you can raise an issue.

**Comment by woyouyitouchenfeijv** [comment]
**version**

- etcd : 
- etcd Version: 3.5.1
- Go OS/Arch: linux/amd64

- apisix
- apisix version 2.12.0

- nginx version: openresty/1.19.3.2
- OpenSSL 1.1.1k  FIPS 25 Mar 2021


Sorry for being late. There's something wrong with what I said before

**step**

1. Create two routes
`{
  "uri": "/code*",
  "name": "code1",
  "methods": [
    "GET",
    "POST",
    "PUT",
    "DELETE",
    "PATCH",
    "HEAD",
    "OPTIONS",
    "CONNECT",
    "TRACE"
  ],
  "upstream": {
    "nodes": [
      {
        "host": "1.1.1.1",
        "port": 8080,
        "weight": 1
      }
    ],
    "timeout": {
      "connect": 6,
      "send": 6,
      "read": 6
    },
    "type": "roundrobin",
    "scheme": "http",
    "pass_host": "pass",
    "keepalive_pool": {
      "idle_timeout": 60,
      "requests": 1000,
      "size": 320
    }
  },
  "status": 1
}`

------------------
`{
  "uri": "/code2*",
  "name": "code2",
  "methods": [
    "GET",
    "POST",
    "PUT",
    "DELETE",
    "PATCH",
    "HEAD",
    "OPTIONS",
    "CONNECT",
    "TRACE"
  ],
  "plugins": {
    "proxy-rewrite": {
      "uri": "/aaa"
    }
  },
  "upstream": {
    "nodes": [
      {
        "host": "1.1.1.1",
        "port": 8081,
        "weight": 1
      }
    ],
    "timeout": {
      "connect": 6,
      "send": 6,
      "read": 6
    },
    "type": "roundrobin",
    "scheme": "http",
    "pass_host": "pass",
    "keepalive_pool": {
      "idle_timeout": 60,
      "requests": 1000,
      "size": 320
    }
  },
  "status": 1
}`


 
2. Direct access apisix:9080/code2

At this time, if you visit apisix:9080/code2, you will jump to 8080/aaa
However, when I close the code proxy of 8080, it will be normal to visit apisix:9080/code2 to 8081/aaa

**Comment by tzssangglass** [comment]
> At this time, if you visit apisix:9080/code2, you will jump to 8080/aaa

my recurrence steps:

1. nginx conf(as upstream)

```nginx
http {
    server {
        listen 8081;

        access_log off;
        location / {
            return 200 "8081";
        }
    }

    server {
        listen 8082;

        access_log off;
        location / {
             return 200 "8082";
        }
    }
}
```

2. create route 1

```shell
curl --location --request PUT 'http://127.0.0.1:9080/apisix/admin/routes/1' \
--header 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' \
--header 'Content-Type: application/json' \
--data-raw '{
    "uri": "/code*",
    "name": "code1",
    "methods": [
        "GET",
        "POST",
        "PUT",
        "DELETE",
        "PATCH",
        "HEAD",
        "OPTIONS",
        "CONNECT",
        "TRACE"
    ],
    "upstream": {
        "nodes": [
            {
                "host": "127.0.0.1",
                "port": 8081,
                "weight": 1
            }
        ],
        "timeout": {
            "connect": 6,
            "send": 6,
            "read": 6
        },
        "type": "roundrobin",
        "scheme": "http",
        "pass_host": "pass",
        "keepalive_pool": {
            "idle_timeout": 60,
            "requests": 1000,
            "size": 320
        }
    },
    "status": 1
}'
```


3. create route 2

```shell
curl --location --request PUT 'http://127.0.0.1:9080/apisix/admin/routes/2' \
--header 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' \
--header 'Content-Type: application/json' \
--data-raw '{
    "uri": "/code2*",
    "name": "code2",
    "methods": [
        "GET",
        "POST",
        "PUT",
        "DELETE",
        "PATCH",
        "HEAD",
        "OPTIONS",
        "CONNECT",
        "TRACE"
    ],
    "plugins": {
        "proxy-rewrite": {
            "uri": "/aaa"
        }
    },
    "upstream": {
        "nodes": [
            {
                "host": "127.0.0.1",
                "port": 8082,
                "weight": 1
            }
        ],
        "timeout": {
            "connect": 6,
            "send": 6,
            "read": 6
        },
        "type": "roundrobin",
        "scheme": "http",
        "pass_host": "pass",
        "keepalive_pool": {
            "idle_timeout": 60,
            "requests": 1000,
            "size": 320
        }
    },
    "status": 1
}'
```

4. send request to APISIX


```shell
$ curl http://127.0.0.1:9080/code2
8082
```

The request is captured by route 2 and sent upstream to port 8082.

I can't reproduce the problem that I said to you.
